File: practica 1/sesion 3/DS_S3_C++Filters/FiltroParteTrabajo.cpp
#include <exception>
using namespace std;

#include "FiltroParteTrabajo.h"
#include "Trabajo.h"
#include "Filtro.h"

void FiltroParteTrabajo::filtrar(Trabajo * unTrabajo) {
	const float extraTrabajo = 0.1;		// Porcentaje extra por trabajo bien hecho

	ParteTrabajo * parte = unTrabajo->getParteTrabajo();

	int horas 			 = parte->getHorasExtra();
	int material 		 = parte->getCostoMaterialesExtra();
	int desplazamiento = parte->getCosteDesplazamiento();

	float mediaParte = ((horas + desplazamiento + material)/3)/10; // dividimos entre 10 para normalizar (0-1)

	float presupuestoAntiguo = unTrabajo->getPresupuesto();
	float nuevoPresupuesto = presupuestoAntiguo + (presupuestoAntiguo*extraTrabajo) * mediaParte;

	/*Al presupuesto antiguo le sumamos como mucho el 10% de su valor anterior (extraTrabajo)
	  la cantidad exacta a sumar vendra determinada por la media del parte, es decir, si la media
	  es un 10, entonces le sumaremos el 10% al completo
	*/

	unTrabajo->setPresupuesto(nuevoPresupuesto);
	// cout << "Filtrando trabajo nuevoPresupuesto " << nuevoPresupuesto << endl;
}

File: practica 3/implementacion dart p3/lib/ModeloSOSEMADO/FiltroParteTrabajo.dart
import 'Trabajo.dart';
import 'Filtro.dart';
import 'ParteTrabajo.dart';

class FiltroParteTrabajo extends Filtro{
  @override
  void filtrar(Trabajo unTrabajo) {
    const double extraTrabajo = 0.1;

    ParteTrabajo parte = unTrabajo.getParteTrabajo();

    int horas = parte.getHorasExtra();
    int material = parte.getCosteMaterialesExtra();
    int desplazamiento = parte.getCosteDesplazamiento();

    double mediaParte = ((horas+material+desplazamiento)/3)/10;

    double presupuestoAntiguo = unTrabajo.getPresupuesto();
    double nuevoPresupuesto = presupuestoAntiguo * (1+extraTrabajo*mediaParte);

    unTrabajo.setPresupuesto(nuevoPresupuesto);
  }
}